sql/memo/rel_props.go

```diff
@@ -257,6 +263,99 @@ func (p *relProps) populateFds() {
 	p.fds = fds
 }
 
+func statsForRel(rel RelExpr) sql.Statistic {
+	var stat sql.Statistic
+	switch rel := rel.(type) {
+	case JoinRel:
+		// different joins use different ways to estimate cardinality of outputs
+		jp := rel.JoinPrivate()
+		left := float64(jp.Left.RelProps.GetStats().RowCount())
+		right := float64(jp.Left.RelProps.GetStats().RowCount())
+
+		switch n := rel.(type) {
+		case *LookupJoin:
+			var card float64
+			if n.Injective {
+				card = n.Left.RelProps.card
+			} else {
+				sel := lookupJoinSelectivity(n.Lookup, n.JoinBase)
+				card = n.Left.RelProps.card * n.Right.RelProps.card * sel
+			}
+			return &stats.Statistic{RowCnt: uint64(card)}
+		case *ConcatJoin:
+			var sel float64
+			for _, l := range n.Concat {
+				lookup := l
+				sel += lookupJoinSelectivity(lookup, n.JoinBase)
+			}
+			card := n.Left.RelProps.card * optimisticJoinSel * sel
+			return &stats.Statistic{RowCnt: uint64(card)}
+		case *LateralJoin:
+			card := n.Left.RelProps.card * n.Right.RelProps.card
+			return &stats.Statistic{RowCnt: uint64(card)}
+		default:
+		}
+
+		switch {
+		case jp.Op.IsPartial():
+			return &stats.Statistic{RowCnt: uint64(left * right * optimisticJoinSel)}
+		case jp.Op.IsLeftOuter():
+			card := math.Max(jp.Left.RelProps.card, optimisticJoinSel*left*right)
+			return &stats.Statistic{RowCnt: uint64(card)}
+		case jp.Op.IsRightOuter():
+			card := math.Max(jp.Right.RelProps.card, optimisticJoinSel*jp.Left.RelProps.card*jp.Right.RelProps.card)
+			return &stats.Statistic{RowCnt: uint64(card)}
+		case jp.Op.IsDegenerate():
+			return &stats.Statistic{RowCnt: uint64(left * right)}
+		default:
+			card := optimisticJoinSel * left * right
+			return &stats.Statistic{RowCnt: uint64(card)}
+		}
+	case *Max1Row:
+		stat = &stats.Statistic{RowCnt: 1}
+
+	case *EmptyTable:
+		stat = &stats.Statistic{RowCnt: 0}
+
+	case *IndexScan:
+		stat = &stats.Statistic{RowCnt: 10}
+
+	case *Values:
+		stat = &stats.Statistic{RowCnt: uint64(len(rel.Table.ExpressionTuples))}
+
+	case *TableFunc:
+		stat = &stats.Statistic{RowCnt: 10}
+
+	case SourceRel:
+		if tn, ok := rel.TableIdNode().(sql.TableNode); ok {
+			if prov := rel.Group().m.StatsProvider(); prov != nil {
+				if card, err := prov.RowCount(rel.Group().m.Ctx, tn.Database().Name(), tn.Name()); err == nil {
+					stat = &stats.Statistic{RowCnt: card}
+					break
+				}
+			}
+		}
+		stat = &stats.Statistic{RowCnt: 1000}
+
+	case *Filter:
+		card := float64(rel.Child.RelProps.GetStats().RowCount()) * .75
```

Comment:
This should probably be outlined into a `filter_selectivity` constant.

Reply:
good idea